### PR TITLE
[flink] Add KafkaCanalEventParser

### PR DESCRIPTION
**Pull request by Alibaba-HZY**
### Purpose

Add canal-json eventParser for kafka canal cdc

### Tests

KafkaEventParserTest

### API and Format

no

### Documentation
no

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.action.cdc.kafka;
+
+import java.util.Arrays;
+import java.util.List;
+
+/** Get precision or scale from string type. */
+public class TypeUtil {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java:25]: What type util? `CDCStringTypeUtil`? Used for what? Add comments? Why not in MysqlTypeUtil?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalValueConvert.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.action.cdc.kafka.canal;
+
+/** Converts some special types such as enum、set、geometry. */
+public class CanalValueConvert {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalValueConvert.java:22]: `CanalFieldParser`?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaEventParserTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+/** Tests for {@link CanalJsonEventParser}. */
+public class KafkaEventParserTest {
+    private static final String CANAL_JSON_EVENT =
+            "{\"data\":[{\"pt\":\"1\",\"_id\":\"1\",\"v1\":\"one\"}],"
+                    + "\"database\":\"paimon_sync_table\",\"es\":1683006706000,\"id\":92,\"isDdl\":false,"
+                    + "\"mysqlType\":{\"pt\":\"INT\",\"_id\":\"INT\",\"v1\":\"VARCHAR(10)\"},\"old\":null,\"pkNames\":[\"_id\"],"
+                    + "\"sql\":\"\",\"sqlType\":{\"pt\":4,\"_id\":4,\"v1\":12},\"table\":\"schema_evolution_1\","
+                    + "\"ts\":1683006706728,\"type\":\"INSERT\"}\n";
+    private static final String DEBEZIUM_JSON_EVENT =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaEventParserTest.java:47]: Maybe we can some test for DDLs?

**Comment by JingsongLi** [comment]
Thanks @Alibaba-HZY , left three minor comments.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalValueConvert.java (hunk 2 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalValueConvert.java:22]: ok

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaEventParserTest.java (hunk 3 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaEventParserTest.java:47]: yes i will add it

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java (hunk 1 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java:25]: yes i  think we can put this in MysqlTypeUtil

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlTypeUtils.java [hunk 4]
```diff
@@ -232,4 +270,110 @@ public static DataType toDataType(
...
+        } else {
+            return typeName;
+        }
+    }
+
+    public static int getPrecision(String typeName) {
+        if (typeName.contains(LEFT_BRACKETS)
+                && typeName.contains(RIGHT_BRACKETS)
+                && isScaleType(typeName)) {
+            int precision =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlTypeUtils.java:340]: Local variable 'precision' is redundant

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlTypeUtils.java [hunk 5]
```diff
@@ -232,4 +270,110 @@ public static DataType toDataType(
...
+
+    public static boolean isSetType(String typeName) {
+        return typeName.toLowerCase().startsWith(SET);
+    }
+
+    /* Get type after the brackets are removed.*/
+    public static String getShortType(String typeName) {
+
+        if (typeName.contains(LEFT_BRACKETS) && typeName.contains(RIGHT_BRACKETS)) {
+            String type =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlTypeUtils.java:327]: be careful about ide check, here: Local variable 'type' is redundant

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                        result =
+                                result.stream()
+                                        .filter(dataField -> !dataField.name().equals(columnName))
+                                        .collect(Collectors.toList());
+                        result.add(new DataField(id, columnName, dataType));
+                        mySqlFieldTypes.put(columnName, columnType);
+                        paimonFieldTypes.put(columnName, MySqlTypeUtils.toDataType(columnType));
+                    }
+                } else if (sqlAlterTableItem instanceof SQLAlterTableDropColumnItem) {
+                    SQLAlterTableDropColumnItem sqlAlterTableDropColumnItem =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:188]: No test cover

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            }
+
+            String oldValue = objectValue.toString();
+            String newValue = oldValue;
+            if (MySqlTypeUtils.isSetType(MySqlTypeUtils.getShortType(mySqlType))) {
+                newValue = CanalFieldParser.convertSet(newValue, mySqlType);
+            } else if (MySqlTypeUtils.isEnumType(MySqlTypeUtils.getShortType(mySqlType))) {
+                newValue = CanalFieldParser.convertEnum(newValue, mySqlType);
+            } else if (MySqlTypeUtils.isGeoType(MySqlTypeUtils.getShortType(mySqlType))) {
+                try {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:320]: No test cover

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+                } catch (Exception e) {
+                    throw new IllegalArgumentException(
+                            String.format("Failed to convert %s to geometry JSON.", oldValue), e);
+                }
+            }
+            resultMap.put(fieldName, newValue);
+        }
+        // generate values of computed columns
+        for (ComputedColumn computedColumn : computedColumns) {
+            resultMap.put(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:334]: No test cover

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+                    computedColumn.columnName(),
+                    computedColumn.eval(resultMap.get(computedColumn.fieldReference())));
+        }
+
+        return resultMap;
+    }
+
+    private Map<String, String> keyCaseInsensitive(Map<String, String> origin) {
+        Map<String, String> keyCaseInsensitive = new HashMap<>();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:343]: No test cover

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**